### ISSUE TITLE
Plotting indicators that use second data source

### ISSUE THREAD

**Issue by happydasch** [verybadsoldier/backtrader_plotting#60]
When backtesting and using more than one data source, indicators applied to second data source contain nan values between single points. This will not render correctly in bokeh, since bokeh assumes, that nan values should create gaps in lines. 

So when backtesting this will result in single not connected points:
in this screenshot I hover above one point:
<img width="1165" alt="Bildschirmfoto 2020-07-05 um 03 09 23" src="https://user-images.githubusercontent.com/4503314/86523510-e4596d00-be6d-11ea-84c2-12a635215765.png">

in live, the line looks this way:
<img width="1188" alt="Bildschirmfoto 2020-07-05 um 03 08 57" src="https://user-images.githubusercontent.com/4503314/86523513-faffc400-be6d-11ea-8896-7f513e93b73d.png">

a suggestion on how to fix this is using numpy interp. So if there are different clocks in use, maybe it would be useful to not fill with nan but interpolate the missing values?

**Comment by happydasch** [comment]
this also applies to plotting lines on data:

on this screenshot two ema lines are missing:

<img width="1192" alt="Bildschirmfoto 2020-07-05 um 03 22 48" src="https://user-images.githubusercontent.com/4503314/86523586-f4258100-be6e-11ea-8a00-acea2ae2201c.png">

**Comment by happydasch** [comment]
Plotting using matplotlib looks this way: 

<img width="904" alt="Bildschirmfoto 2020-07-05 um 03 32 27" src="https://user-images.githubusercontent.com/4503314/86523704-3e5b3200-be70-11ea-913e-7cae4fc9faff.png">

**Comment by happydasch** [comment]
I dig a little bit into this. The reason for this behavior is the use of a single cds. When using a different cds for every object to plot, which contains just the points to show, the line will show. If in the y values for a line is None -> line will be plotted at 0, if float('nan') in y values, line will discontinue at this point.

So to overcome this, there may be 2 possibilities:
-use a cds for every figure (or even line)
-fill the missing values with interpolated data (this will lead to add checks if line is being plotted as a line, as a marker, etc., this would introduce a source for some problems, when manipulating the data source)

I will try to add some functionality when the line is being plotted to create a cds for every figure from the generated data in get_strategy_data, so every figure / line can take its data from this, remove the nan values and just put this into its own cds)

**Comment by happydasch** [comment]
This will show the problem:


```
from bokeh.plotting import figure, output_file, show

output_file("line.html")

p = figure(plot_width=400, plot_height=400)

# add a line renderer
p.line([1, 2, 3, 4, 5], [6, 7, float('nan'), 4, 5], line_width=2)
#p.line([1, 2, 3, 4, 5], [6, 7, None, 4, 5], line_width=2)

show(p)
```

**Comment by verybadsoldier** [comment]
Thank you, good point.

**Comment by happydasch** [comment]
So i intend to fix it that way:
-remove any authority from cds (checking for data in cds) -> will use only the generated data from get_strategy_data
-when data is generated, don't push it into cds but dispatch it somehow to every line / marker / etc, that is being plotted
-there they can decide if the last row and the corresponding col is interesting for its cds (if there is a value, put it in the cds of the line - index and value will be put into that cds, if not, just skip it)

**Comment by verybadsoldier** [comment]
> When backtesting and using more than one data source, indicators applied to second data source contain nan values between single points. This will not render correctly in bokeh, since bokeh assumes, that nan values should create gaps in lines.

Not sure why indicators of second data sources contain NaN values in the first place though. Do you have an example to inspect?

**Comment by happydasch** [comment]
Its basically this method:

https://github.com/verybadsoldier/backtrader_plotting/blob/01e1d8ed90604cb7a5ed84c57fdfbb3a8ceeb872/backtrader_plotting/utils.py#L50

def convert_by_line_clock(line, line_clk, new_clk)

The data of the indicator, which is generated on larger timeframe (5 minutes(data0) -> 1 hour (data1)) will be spread out over the strategy clock index. So there will be 11 nans, 1 value, 11 nans, 1 value

**Comment by verybadsoldier** [comment]
So this is exclusively about two data sources that do no share the same timeframe?

**Comment by happydasch** [comment]
This basically should affect everything with more than one data source. Even plotting data as a line. I will provide a sample strategy in a few minutes.

**Comment by happydasch** [comment]
Here is a example of the problem, I set style="line" -> no line of data1, set it to style="bar" bars are plotted.

in data1 there should be an ema line, when you hover above the second data, the points are there.

[line_render.zip](https://github.com/verybadsoldier/backtrader_plotting/files/4874838/line_render.zip)

**Comment by happydasch** [comment]
so basically for every column in generated df, there should be a cds containing the index and value without nan values. for cds that are used to create the x axis, there should also be a datetime (basically data sources with/without volume)

**Comment by verybadsoldier** [comment]
Thanks, but the example you posted actually is showing two data sources that use _different_ timeframes. I understand the problem with different timeframes. Did it happen for you also when using same timeframes?

**Comment by happydasch** [comment]
No, it won't happen with a single timeframe, since there will be most probably no gaps. this is also why when using live client, the lines will be rendered correctly. it will appear on the generated backtest html file when using different timeframes.

**Comment by happydasch** [comment]
It took me some time, but I fixed that issue. The fix is actually very easy. I did the following:

-FigurePage now has a fill_data_from_df(df) method
-FigurePage has a cds with all data available (for hover, for x_axis) this is also useful, when working with the liveclient to fill with new data
-FigureEnvelope has a fill_data_from_df(df) method
-Every figure now knows what columns it is showing by collecting the columns to show in self._cds_cols, by collecting them when plot is called
-new function set_cds_columns_from_df which will take the columns from the df which are required and set the provided cds with just this columns (it will set them up and fill with data from df)
-The FigurePage (and all FigureEnvelopes) will now be filled when _blueprint_strategy is called by generating the df for tradingdomain if any is set (I removed in my fork the fill_data param from plot, I will work on the liveclient later) 

So it basically goes that way:
-Bokeh. _blueprint_strategy is called -> all figures are created (getting the FigurePage.cds as a param) for current FigurePage -> After everything is done FigurePage.fill_data_from_df(df) is called -> this provides all figures in that page with the df -> The figures itself take the columns they use from the df and remove all columns with no data (this happens in set_cds_columns_from_df)

You could add this functionality and move the fill_data_from_df call to plot, but you would need to ensure, that it is at least one time called before plotting anything, since the cds of the figure page and figures are created when calling this. The liveclient is not yet updated but I committed the fix.

**Comment by happydasch** [comment]
https://github.com/verybadsoldier/backtrader_plotting/issues/60#issuecomment-653830093

After the fix the secondary data is plotted this way:

<img width="1151" alt="Bildschirmfoto 2020-07-06 um 01 02 54" src="https://user-images.githubusercontent.com/4503314/86544116-99e6f780-bf24-11ea-8573-adcf0cd3bbfc.png">

**Comment by happydasch** [comment]
I have now also committed the liveclient to work with multiple cds, patching works, streaming works also. The patching of replaydata seems to work also fine.
